# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/nsFrameTraversal.h
# Commit: 92148b810f22
# Full Hash: 92148b810f228e4e9e84255682fe888a5bee07d5
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-12-24 09:09:23
# Regressor Bug: 779684
# File Overlap Count: 1
# Description:
#   Bug 779684 - part 2: Get rid of `nsFrameTraversal` and `nsIFrameTraversal` r=emilio
#   
#   The class is used only for creating `nsFrameIterator`, but it's unnecessary
#   so that we can get rid of it and its interface.
#   
# ==============================================================================

diff -r d4411b38cae7 -r 92148b810f22 layout/base/nsFrameTraversal.h
--- a/layout/base/nsFrameTraversal.h	Sun Dec 24 06:01:19 2023 +0000
+++ b/layout/base/nsFrameTraversal.h	Sun Dec 24 06:01:19 2023 +0000
@@ -8,9 +8,10 @@
 #define NSFRAMETRAVERSAL_H
 
 #include "mozilla/Attributes.h"
-#include "nsIFrameTraversal.h"
+#include "nsISupportsImpl.h"
 
 class nsIFrame;
+class nsPresContext;
 
 class nsFrameIterator {
   NS_INLINE_DECL_REFCOUNTING(nsFrameIterator)
@@ -33,11 +34,23 @@
     return CurrentItem();
   };
 
-  nsFrameIterator(nsPresContext* aPresContext, nsIFrame* aStart,
-                  nsIteratorType aType, bool aLockScroll, bool aFollowOOFs,
-                  bool aSkipPopupChecks, nsIFrame* aLimiter);
+  enum class Type : uint8_t {
+    // only leaf nodes
+    Leaf,
+    // "open tag" order
+    PreOrder,
+    // "close tag" order
+    PostOrder,
+  };
+  static already_AddRefed<nsFrameIterator> Create(
+      nsPresContext* aPresContext, nsIFrame* aStart, Type aType, bool aVisual,
+      bool aLockInScrollView, bool aFollowOOFs, bool aSkipPopupChecks,
+      nsIFrame* aLimiter = nullptr);
 
  protected:
+  nsFrameIterator(nsPresContext* aPresContext, nsIFrame* aStart, Type aType,
+                  bool aLockInScrollView, bool aFollowOOFs,
+                  bool aSkipPopupChecks, nsIFrame* aLimiter);
   virtual ~nsFrameIterator() = default;
 
   void SetCurrent(nsIFrame* aFrame) { mCurrent = aFrame; }
@@ -102,7 +115,7 @@
   const bool mLockScroll;
   const bool mFollowOOFs;
   const bool mSkipPopupChecks;
-  const nsIteratorType mType;
+  const Type mType;
 
  private:
   nsIFrame* const mStart;
@@ -112,30 +125,4 @@
   int8_t mOffEdge;  // 0= no -1 to far prev, 1 to far next;
 };
 
-nsresult NS_NewFrameTraversal(nsFrameIterator** aEnumerator,
-                              nsPresContext* aPresContext, nsIFrame* aStart,
-                              nsIteratorType aType, bool aVisual,
-                              bool aLockInScrollView, bool aFollowOOFs,
-                              bool aSkipPopupChecks,
-                              nsIFrame* aLimiter = nullptr);
-
-nsresult NS_CreateFrameTraversal(nsIFrameTraversal** aResult);
-
-class nsFrameTraversal final : public nsIFrameTraversal {
- public:
-  nsFrameTraversal();
-
-  NS_DECL_ISUPPORTS
-
-  NS_IMETHOD NewFrameTraversal(nsFrameIterator** aEnumerator,
-                               nsPresContext* aPresContext, nsIFrame* aStart,
-                               int32_t aType, bool aVisual,
-                               bool aLockInScrollView, bool aFollowOOFs,
-                               bool aSkipPopupChecks,
-                               nsIFrame* aLimiter = nullptr) override;
-
- protected:
-  virtual ~nsFrameTraversal();
-};
-
 #endif  // NSFRAMETRAVERSAL_H